applications/asset_tracker_v2/src/modules/location_module.c

```diff
@@ -0,0 +1,634 @@
+			.module.util = *event
+		};
+
+		message_handler(&msg);
+	}
+
+	if (is_location_module_event(aeh)) {
+		struct location_module_event *event = cast_location_module_event(aeh);
+		struct location_msg_data msg = {
+			.module.location = *event
+		};
+
+		message_handler(&msg);
+	}
+
+	if (is_modem_module_event(aeh)) {
+		struct modem_module_event *event = cast_modem_module_event(aeh);
+		struct location_msg_data msg = {
+			.module.modem = *event
+		};
+
+		message_handler(&msg);
+	}
+	return false;
+}
+
+/* Static module functions. */
+static void inactivity_timeout_handler(struct k_timer *timer_id)
+{
+	/* Simulate a sleep event from GNSS with older MFWs. */
+	int event = NRF_MODEM_GNSS_EVT_SLEEP_AFTER_TIMEOUT;
+
+	k_msgq_put(&event_msgq, &event, K_NO_WAIT);
+}
+
+static K_TIMER_DEFINE(inactivity_timer, inactivity_timeout_handler, NULL);
+
+static void timeout_send(void)
+{
+	struct location_module_event *location_module_event = new_location_module_event();
+
+	location_module_event->data.location.search_time = stats.search_time;
+	location_module_event->type = LOCATION_MODULE_EVT_TIMEOUT;
+
+	APP_EVENT_SUBMIT(location_module_event);
+}
+
+static void data_send_pvt(void)
+{
+	struct location_module_event *location_module_event = new_location_module_event();
+
+	location_module_event->data.location.pvt.longitude = pvt_data.longitude;
+	location_module_event->data.location.pvt.latitude = pvt_data.latitude;
+	location_module_event->data.location.pvt.altitude = pvt_data.altitude;
+	location_module_event->data.location.pvt.accuracy = pvt_data.accuracy;
+	location_module_event->data.location.pvt.speed = pvt_data.speed;
+	location_module_event->data.location.pvt.heading = pvt_data.heading;
+	location_module_event->data.location.timestamp = k_uptime_get();
+	location_module_event->type = LOCATION_MODULE_EVT_GNSS_DATA_READY;
+
+	location_module_event->data.location.satellites_tracked = stats.satellites_tracked;
+	location_module_event->data.location.search_time =
+		(uint32_t)(location_module_event->data.location.timestamp - stats.start_uptime);
+
+	APP_EVENT_SUBMIT(location_module_event);
+}
+
+static void search_start(void)
+{
+	int err;
+	struct location_config config;
+	enum location_method methods[2];
+	int methods_count = 0;
+	int methods_index_gnss = -1;
+	int methods_index_cellular = -1;
+
+	if (copy_cfg.no_data.neighbor_cell && copy_cfg.no_data.gnss) {
+		SEND_EVENT(location, LOCATION_MODULE_EVT_DATA_NOT_READY);
+		LOG_ERR("Both GNSS and cellular configured off");
+		return;
+	}
+
+	if (!copy_cfg.no_data.gnss) {
+		methods[methods_count] = LOCATION_METHOD_GNSS;
+		methods_index_gnss = methods_count;
+		methods_count++;
+	}
+
+	if (!copy_cfg.no_data.neighbor_cell) {
+		methods[methods_count] = LOCATION_METHOD_CELLULAR;
+		methods_index_cellular = methods_count;
+		methods_count++;
+	}
+
+	if (methods_count == 0) {
+		LOG_WRN("No location methods selected");
+		return;
+	}
+
+	location_config_defaults_set(&config, methods_count, methods);
```

Comment:
This uses the default 120 second timeout for GNSS. I think that's too much with the default sampling interval, which is also 120 seconds. It currently takes so long for GNSS to timeout (when GNSS can't get a fix), that data from the location module is not ready when data is sent to the cloud. Thus, every second batch send does not have location data at all.

I think that the goal is to get some kind of location on each sampling round. Because of this GNSS should timeout in time to do neighbor cell measurements before data is going to be sent to the cloud. With 120s sampling interval, something like 90s GNSS timeout should be sufficient.

Reply:
As we've discussed in the office, this is complicated and timeouts will have some flaws always. Reducing the timeout doesn't guarantee that we get cellular position for every sampling with 120s interval because 90s GNSS timeout starts only when RRC is not connected and if we don't have A-GPS we need to get to PSM also.

So the longer this is, the more chances we have to get a fix at some point when few intervals have gone.
And this would need to be shorter for us to be able to fail and fall back to cellular positioning.

I'll leave this pending for a while as resolution to the following comment might have an effect on this:
https://github.com/nrfconnect/sdk-nrf/pull/8631#discussion_r1018984894

Reply:
I've changed GNSS timeout to 90s. We can update this later if we come up with a better value.